sklearn/preprocessing/_encoders.py

```diff
@@ -41,18 +41,38 @@ def _check_X(self, X):
           not do that)
 
         """
+        if hasattr(X, 'iloc') and getattr(X, 'ndim', 0) == 2:
+            # only if pandas dataframe (not an array, sequence etc...)
+            # do not check full X,
+            # it should be checked column by column (using _check_X_feature())
+            return X
+
         X_temp = check_array(X, dtype=None)
         if not hasattr(X, 'dtype') and np.issubdtype(X_temp.dtype, np.str_):
             X = check_array(X, dtype=np.object)
         else:
             X = X_temp
 
-        if X.dtype == np.dtype('object'):
+        return X
+
+    def _check_X_feature(self, Xi):
+        Xi = np.asarray(Xi)
+        Xi = check_array(Xi, ensure_2d=False, dtype=None)
```

Comment:
add force_all_finite=False here. That removes any O(n) processing from check_array, so duplication is no big deal.